File: Algolab-Algorithm(2022)/NQueen/NQueen.cpp
#include <iostream>

using namespace std;

int num;
int answer = 0;
int col[16];

void NQueen(int row);
int isPromising(int row);
int main()
{
    int numTestCase;
    cin >> numTestCase;
    for (int i = 0; i < numTestCase; i++)
    {
        cin >> num;
        for (int j = 0; j < num; j++)
        {
            col[0] = j;
            NQueen(0);
        }
        answer = 0;
    }
}

void NQueen(int row)
{
    if (answer == 1)
    {
        return;
    }
    if (isPromising(row))
    {
        if (row == num - 1)
        {
            for (int i = 0; i < num; i++)
            {
                cout << col[i] + 1 << " ";
            }
            cout << endl;
            answer = 1;
        }
        else
        {
            for (int i = 0; i < num; i++)
            {
                col[row + 1] = i;
                NQueen(row + 1);
            }
        }
    }
}

int isPromising(int row)
{
    int k = 0;
    int promising = 1;
    while (k < row && promising)
    {
        if (col[row] == col[k] || abs(col[row] - col[k]) == row - k)
        {
            promising = 0;
        }
        k++;
    }
    return promising;
}

File: Baekjoon Algorithm/문제/백트래킹/N-Queen/N-Queen.py
import sys

input = sys.stdin.readline

N = int(input())

answer = 0
row = [0] * N

def is_promising(x):
    for i in range(x):
        if row[x] == row[i] or abs(row[x]-row[i]) == abs(x-i):
            return False
    return True
    

def N_Queen(x):
    global answer
    if x == N:
        answer += 1
        return
    else:
        for i in range(N):
            row[x] = i
            if is_promising(x):
                N_Queen(x+1)

N_Queen(0)
print(answer)